core/shared/utils/bh_common.c

```diff
@@ -167,15 +167,39 @@ wa_strdup(const char *s)
 }
 
 #if WASM_ENABLE_WAMR_COMPILER != 0 || WASM_ENABLE_JIT != 0
+/* need to make sure that The `argv[]` must be terminated by a NULL pointer. */
 int
-bh_system(const char *cmd)
+bh_system(const char *pathname, char *const argv[], int argc)
 {
     int ret;
+    /* no environment variables */
+    char *const envp[] = { NULL };
+
+    if (pathname == NULL) {
+        return -1;
+    }
+
+    if (argc > 0) {
+        if (argv == NULL) {
+            return -1;
+        }
+
+        /* The `argv[]` must be terminated by a NULL pointer. */
+        if (argv[argc - 1] != NULL) {
+            return -1;
+        }
+    }
 
 #if !(defined(_WIN32) || defined(_WIN32_))
-    ret = system(cmd);
+    ret = execve(pathname, argv, envp);
```

Comment:
does this work at all? i suppose you need fork+execve+wait.

Reply:
> execve() executes the program referred to by pathname.  This causes the program that is currently being run by the calling process to be replaced with a new program, with newly initialized stack, heap, and (initialized and uninitialized) data segments.

IMU, in our scenarios, there is no need to fork a new process and run commands. The default behavior is sufficient

Reply:
i don't think so. wamrc has other things to do after invoking these external commands.
have you tested?

Reply:
Yes, I used commands like `WMRC_LLC_COMPILER=/SDF/051_llvm_releases/clang+llvm-15.0.6-x86_64-linux-gnu-ubuntu-18.04/bin/clang WAMRC_LLC_FLAGS="-mcmodel=medium" /workspaces/wasm-micro-runtime/wamr-compiler/build/wamrc -o xyz.aot ./xyz.wasm`.

Reply:
I‘m also curious about how this works. The man page says:
> execve() does not return on success, and the text, initialized data, uninitialized data (bss),  and  stack  of
       the calling process are overwritten according to the contents of the newly loaded program.

That means the later instructions after execve will not be executed.

Reply:
Thanks for reminder. Now I see both of your points.